crypto/zk/artifacts/artifacts.go

```diff
@@ -0,0 +1,175 @@
+	by the vochain]
+	- witness.wasm: (medium size, linear with circuit size) [needed by the user]
+- Examples:
+	- zkcensusproof/stage/1024/circuit.zkey
+	- zkcensusproof/dev/128/witness.wasm
+
+# Files hash
+=============
+Each file is hashed with sha256, which output is publicly known. The hash
+values of each file are hardcoded in the VochainGenesis.
+
+*/
+
+const (
+	// FilenameWitness defines the name of the file of the WitnessCalculator
+	FilenameWitness string = "witness.wasm"
+	// FilenameZkey defines the name of the file of the circom ZKey
+	FilenameZkey string = "circuit.zkey"
+	// FilenameVk defines the name of the VerificationKey.json
+	FilenameVk string = "verificationKey.json"
+)
+
+// CircuitsConfig defines the configuration of the files to be downloaded
+type CircuitsConfig struct {
+	// URL defines the URL from where to download the files
+	URL string
+	// CircuitsPath defines the path from where the files are downloaded
+	CircuitsPath string
+	// LocalDir defines in which directory will be the files
+	// downloaded, under that directory it will follow the CircuitsPath
+	// directories structure
+	LocalDir string
+
+	// WitnessHash contains the expected hash for the file filenameWitness
+	WitnessHash []byte // witness.wasm
+	// ZkeyHash contains the expected hash for the file filenameZkey
+	ZkeyHash []byte // circuit.zkey
+	// VkHash contains the expected hash for the file filenameVk
+	VkHash []byte // verificationKey.json
+}
+
+// DownloadCircuitFiles will download the circuits in the specified path,
+// checking the expected sha256 hash of each file. If the files already exist
+// in the path, it will not download them again but it will check the hash of
+// the existing files.
+func DownloadCircuitFiles(c CircuitsConfig) error {
+	// check if files already exist, if files already exist, check hashes
+	err := checkHashes(c)
+	if err == nil {
+		// files already exist, and match the expected hashes
+		return nil
+	}
+	if err != nil && !strings.Contains(err.Error(), "no such file or directory") {
+		return err
+	}
+
+	// at this point, or files do not exist, or exist with a not expected
+	// hash
+	return downloadFiles(c)
+}
+
+func checkHashes(c CircuitsConfig) error {
+	err := checkHash(c.LocalDir+c.CircuitsPath+FilenameWitness, c.WitnessHash)
+	if err != nil {
+		return err
+	}
+
+	err = checkHash(c.LocalDir+c.CircuitsPath+FilenameZkey, c.ZkeyHash)
+	if err != nil {
+		return err
+	}
+
+	err = checkHash(c.LocalDir+c.CircuitsPath+FilenameVk, c.VkHash)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkHash(path string, expected []byte) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return err
+	}
+	hash := h.Sum(nil)
+
+	if !bytes.Equal(hash[:], expected) {
+		return fmt.Errorf("path: %s, expected hash: %s, computed hash: %s",
+			path, hex.EncodeToString(expected), hex.EncodeToString(hash[:]))
```

Comment:
```suggestion
		return fmt.Errorf("path: %s, expected hash: %x, computed hash: %x",
			path, expected, hash[:])
```
Did you know about this?